platform/mellanox/install-pending-fw.py

```diff
@@ -0,0 +1,97 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+import os
+
+from collections import namedtuple
+
+from fwutil.lib import ComponentStatusProvider, PlatformComponentsParser
+from sonic_platform.component import ComponentCPLD, MPFAManager
+
+# Globals
+FW_STATUS_SCHEDULED = "scheduled"
+CPLD_FLAG = False
+
+# Init platform chassis helper classes
+csp = ComponentStatusProvider()
+pcp = PlatformComponentsParser(csp.is_modular_chassis())
+
+# Parse update status file
+update_status = csp.read_au_status_file_if_exists()
+
+if update_status is None:
+    exit(0)
+
+# Parse platform components file
+try:
+    pcp.parse_platform_components()
+except Exception as e:
+    print("Error parsing platform components. Firmware update failed: {}".format(str(e)))
+    print("System will reboot in 10 seconds please fix issue and run update command again.")
+    time.sleep(10)
+    exit(-1)
+
+# Iterate each component in the status file
+comp_install = []
+files = []
+
+for boot_type, components in update_status.items():
+    for comp in components:
+
+        # Skip if fw isn't scheduled for install at reboot
+        if comp["info"] != FW_STATUS_SCHEDULED: continue
+
+        # Get component object and target firmware file
+        key = comp["comp"]
+        comp_path = key.split("/")
+
+        if len(comp_path) == 3:
+            # Module component
+            _, parent_name, comp_name = comp_path
+            fw_file = pcp.module_component_map[parent_name][comp_name]["firmware"]
+            component = csp.module_component_map[parent_name][comp_name]
+        else: 
+            # Chassis component
+            parent_name, comp_name = comp_path
+            fw_file = pcp.chassis_component_map[parent_name][comp_name]["firmware"]
+            component = csp.chassis_component_map[parent_name][comp_name]
+
+        # Install firmware. If CPLD flag to be installed last due to force reboot during refresh
+        if type(component) == ComponentCPLD:
+            if CPLD_FLAG:
+                # Only need one refresh
+                continue
+            CPLD_FLAG = True
+            mpfa = MPFAManager(fw_file)
+            mpfa.extract()
+            if not mpfa.get_metadata().has_option('firmware', 'refresh'):
+                raise RuntimeError("Failed to get {} refresh firmware".format(self.name))
+            refresh_firmware = mpfa.get_metadata().get('firmware', 'refresh')
+            comp_install = comp_install + [component]
+            files = files + [os.path.join(mpfa.get_path(), refresh_firmware)]
+        else:
+            comp_install = [component] + comp_install
+            files = [fw_file] + files
+
+# Do install
+for i, c in enumerate(comp_install):
+    if type(c) == ComponentCPLD:
+        c.install_firmware(files[i])
```

Comment:
@alexrallen how about exceptions handling for FW installation/upgrade?

Reply:
Hrm, python exceptions? This is directly before the reboot so I'm not sure how much any type of exception handling is going to help us here. 

I believe the exceptions should be handled at the implementation of `install_firmware` as necessary (which I believe is the case now). This install call will return True / False depending on the success but there isn't anything we can really act on at this point in the process. The user is going to have to let the switch reboot and check the status of the firmware.

Reply:
@alexrallen there could be a backtrace so the output will be messed.

Please have a look on how it is done in `fwutil`:
```python
https://github.com/Azure/sonic-utilities/blob/master/fwutil/main.py#L209

    try:
        click.echo("Installing firmware:")
        click.echo(TAB + fw_path)
        log_helper.log_fw_install_start(component_path, fw_path)
        status = component.install_firmware(fw_path)
        log_helper.log_fw_install_end(component_path, fw_path, status)
    except KeyboardInterrupt:
        log_helper.log_fw_install_end(component_path, fw_path, False, "Keyboard interrupt")
        raise
    except Exception as e:
        log_helper.log_fw_install_end(component_path, fw_path, False, e)
        cli_abort(ctx, str(e))
```

Reply:
If the file system is locked then at this stage we will not have write access to a log. 

I am fine try-catching this and dumping an error to the terminal though if that is okay with you.

Reply:
@nazariig I have added a commit, please check.